src/core/job.c

```diff
@@ -1332,6 +1333,56 @@ int job_deserialize(Job *j, FILE *f) {
         return 0;
 }
 
+static int job_unit_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_free_ char *p = NULL;
+        Job *j = ASSERT_PTR(userdata);
+
+        assert(ret);
+
+        p = unit_dbus_path(j->unit);
+        if (!p)
+                return -ENOMEM;
+
+        return sd_json_buildo(ret,
+                        SD_JSON_BUILD_PAIR_STRING("id", j->unit->id),
+                        SD_JSON_BUILD_PAIR_STRING("path", p));
+}
+
+static int activation_details_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_strv_free_ char **pairs = NULL;
+        ActivationDetails *activation_details = userdata;
+        int r;
+
+        assert(ret);
+
+        r = activation_details_append_pair(activation_details, &pairs);
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH_PAIR(key, value, pairs) {
+                r = sd_json_variant_append_arraybo(&v,
+                                SD_JSON_BUILD_PAIR_STRING("key", *key),
+                                SD_JSON_BUILD_PAIR_STRING("value", *value));
```

Comment:
shouldn't this be:

`JSON_BUILD_PAIR_STRING(*key, *value)`

?

Reply:
Then it becomes completely untypable in introspection?